sw/device/lib/testing/mock_mmio.h

```diff
@@ -0,0 +1,78 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_TESTING_MOCK_MMIO_H_
+#define OPENTITAN_SW_DEVICE_LIB_TESTING_MOCK_MMIO_H_
+
+#include <stdint.h>
+
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+
+namespace mock_mmio {
+/**
+ * A MockDevice represents a mock implementation of an MMIO device.
+ *
+ * MockDevice provides two mockable member functions, representing a read and a
+ * write at a particular offset from the base address. This class can be
+ * converted into a |reg32_t| value, which, when used in |mmio.h| functions like
+ * |reg32_read()|, will map to the appropriate mock member function calls.
+ *
+ * To maintain sequencing, |ReadInternal()| and |WriteInternal()| should not be
+ * |EXPECT_CALL|'ed directly; instead, |EXPECT_READ| and |EXPECT_WRITE| should
+ * be used, instead.
+ *
+ * To use this class, |-DMOCK_MMIO| must be enabled in all translation units
+ * using |mmio.h|.
+ */
+class MockDevice {
+ public:
+  MockDevice() = default;
+
+  MockDevice(const MockDevice &) = delete;
+  MockDevice &operator=(const MockDevice &) = delete;
+  MockDevice(MockDevice &&) = delete;
```

Comment:
IIRC, move constructor is implicitly deleted here since a copy constructor is declared but I think this makes the intent clearer.

Reply:
Google C++ requires that you explicitly delete everything because no one ever remembers the Rule of Five correctly.